src/backend/distributed/shared_library_init.c

```diff
@@ -799,10 +799,10 @@ RegisterCitusConfigVariables(void)
 
 	DefineCustomBoolVariable(
 		"citus.enable_alter_role_propagation",
-		gettext_noop("Enables propagating role passwords statements to workers"),
+		gettext_noop("Enables propagating ALTER ROLE statements to workers"),
 		NULL,
 		&EnableAlterRolePropagation,
-		false,
+		true,
```

Comment:
One thing that scares me a little is that we'll now auto-propagate things that the user originally only meant to set on the coordinator, e.g.:
```sql
ALTER ROLE citus SET enable_hashagg TO off;
```

That could (after a while) lead to some dramatic, unexpected performance regressions when upgrading existing clusters.

One alternative could be that we require some prefix if you want the GUC to propagate, e.g.:
```sql
ALTER ROLE citus SET workers.enable_hashagg TO off;
```

That way, the behaviour remains consistent across versions and we preserve the ability to set separate configs on coordinator and worker. 

I'm guessing this is not a big implementation change. What do you think?

Reply:
I think this way of propagating settings is nice! Don't know how much work it would be to change the alter role set for this.

Would have been great if we though of this a month ago :)

Implementation is not much work I guess, but tests are messy so might cause some issues. I will stack a PR on top of this one to see the feasibility.

Reply:
Thinking more about this, I think this distinction should not be made on the user level, but on the node level.

disabling `hashagg` on the coordinator and not on the workers is specific to the topology of the system, not to the user using the system.

I do like the idea though, and maybe we should holistically figure out how to solve this. Holistically because there are multiple level where postgres allowes you to apply settings:
 - `ALTER SYSTEM`
 - `ALTER ROLE ALL`
 - `ALTER ROLE ... IN DATABASE`
 - `ALTER ROLE ALL IN DATABASE`

Citus would add an extra dimension to _all_ settings where we either want them
 - `globally`; the whole cluster
 - `coordinator`; on the node coordinating the query
 - `worker`; the node executing the shard query on top of the data

Lastly it interferes with `MX` where workers would both be coordinating the query, and executing the `worker` part. Add on top local execution :).

Reply:
> Holistically because there are multiple level where postgres allowes you to apply settings:

isn't the workers prefix defining precisely that?

Basically, if I run the following on the coordinator:
```sql
ALTER ROLE workers.enable_hashagg TO off;
```

then we send the following to workers:
```sql
ALTER ROLE citus SET enable_hashagg TO off;
```

We could add another prefix at a later time, e.g.:

```sql
ALTER ROLE citus SET global.enable_hashagg TO off
```

which also runs `ALTER ROLE citus SET enable_hashagg TO off` on the coordinator, but that seems less urgent.

Reply:
This doesn't cover MX where the worker could also be a coordinator in the sense of disabling hashagg on the coordinator. Hence I think we need to have a good specification. Not a last minute try in the release cycle.

Reply:
I guess you're saying that there should be a difference between settings when handling distributed queries and local queries, but I don't think such a distinction is really useful. Users vs. databases vs. nodes are clearly defined session-specific context.

There is no precedent for using different settings for different types of queries in the same context in PostgreSQL, except maybe within function calls.

Reply:
conclusion of an in person chat:

design goals we want to keep
 - in an ideal case we would have all settings synchronised during query execution

reality we live in
 - we have advised often to `ALTER ROLE ... SET enable_hashagg TO off` to prevent OOM on the coordinator
    - ^ This will be solved with pg13 where disk based hash aggs are a thing

In a future we might want to experiment with scoped (`worker`/`coordinator`) settings, but this is a broad task. This should work irregardless of how the setting is set, eg. `SET LOCAL ...`, `SET ...`, `ALTER ROLE ... SET ...`.

Our short term goals:
 - Don't break customers that have `enable_hashagg` set to `off` on the coordinator but not on the workers.

To not block the future too much and also not break existing customers we will introduce a separate guc for `ALTER ROLE ... SET ...` that defaults to `on`. This allows our managed solutions to disable the guc during an upgrade of the cluster. For new clusters that have disk based hash agg support we can decide to have the `alter role set propagation guc` to on to achieve our ideal goal outlined above.